src/dynam.jl

```diff
@@ -0,0 +1,207 @@
+using Catlab
+using Catlab.CategoricalAlgebra
+using Catlab.Theories
+import Catlab.Theories: id, dom, codom, compose
+
+"""    VectorField{F}
+
+A dynamical system over a Euclidean Space on `n` dimensions. 
+Use `VectorField{Function}` to disable specialization on the vector field function.
+"""
+struct VectorField{F}
+    domain::FinSet # Rⁿ
+    field::F  # maps Rⁿ to Tangents(Rⁿ)
+end
+
+domain(U::VectorField) = U.domain
+tanspace(U::VectorField) = U.domain
+vfield(U::VectorField) = U.field
+
+(U::VectorField)(u) = begin
+    FinSet(length(u)) == domain(U) || error("Domain Mismatch: state vector has length $(length(u)) was expecting $(domain(U))")
+    vfield(U)(u)
+end
+
+"""    VectorFieldHom
+
+The structure for storing a homomorphism between VectorField objects. The key axiom is that
+
+    restrict(f, simulate(f, u) - f.u(u)) == 0:domain(f.v)
+
+this means that you can see a version of v inside u by applying the `Pullback(f)` map to restrict states of u to a state of v.
+Then apply the vector field for v to get a tangent vector for v. Then you can apply the `Pushforward(f)` map to get a tangent vector for u.
+
+    v::VectorField
+    u::VectorField
+    f::FinFunction
+   
+    Pullback(f):    domain(u) → domain(v)
+    Pushforward(f): tan(v) → tan(u)
+
+"""
+struct VectorFieldHom
+    v::VectorField
+    u::VectorField
+    f::FinFunction
+end
+
+dom(ϕ::VectorFieldHom) = ϕ.v
+codom(ϕ::VectorFieldHom) = ϕ.u
+
+id(V::VectorField) = VectorFieldHom(V, id(domain(V)))
+compose(ϕ::VectorFieldHom, γ::VectorFieldHom) = begin
+    VectorFieldHom(dom(ϕ),
+                codom(γ),
+                ϕ.f ⋅ γ.f)
+end
+
+pullback(f::FinFunction) = u -> u[collect(f)]
+pushforward(f::FinFunction) = u̇ -> map(codom(f)) do i
+    sum(u̇[j] for j in preimage(f, i);init=0.0)
+end
```

Comment:
Are these two methods used anywhere? Also, looks like you define `pullback` here but `restrict` later on. Probably best to just use one of those words.

And if you use `pullback`, add `import Catlab.CategoricalAlgebra: pullback` to the top.

Reply:
Kris and I were discussing options for pullback because it is so confusing with the pullback of a cospan. I think I can get rid of these functions because I decided to make them a struct containing f with `Pullback` and `Pushforward`. Restrict by precomposition seems more clear to me than pullback in this setting, but is there a different official term?

Reply:
It's an unfortunate fact of life that "pullback" is used in math two different, incompatible ways. If you want to avoid that, I think "restriction" is fine since it just mildly generalizes the usual notion of restricting a function (precomposing with an inclusion map).